### DISCUSSION TITLE
blog/avoiding-use-effect-with-callback-refs

### DISCUSSION THREAD

**Discussion by giscus** [TkDodo/blog-comments#83] [Announcements]
# blog/avoiding-use-effect-with-callback-refs

Interacting with DOM nodes doesn't necessarily need useEffect

https://tkdodo.eu/blog/avoiding-use-effect-with-callback-refs

**Comment by damianricobelli** [comment]
Incredible explanation. I never thought of that before, thank you so much for sharing your knowledge! 🚀

**Comment by ghana7989** [comment]
Wow thanks

**Comment by LookRain** [comment]
Great writeup! This pattern looks really neat and super useful, thanks!
```jsx
const ref = React.useCallback((node) => {
  node?.focus()
}, [])

return <input ref={ref} defaultValue="Hello world" />
```

**Comment by eriicafes** [comment]
Thanks for this, I’ve always been scared of callbackRefs until today

**Comment by lindellcarternyc** [comment]
Bravo! Great explanation on how to use this library in an efficient way. 

Sometimes I feel like we see one way of implementing a common problem and glob onto it. This provides a seemingly stable and safe alternative!

Cheers🎉

**Comment by CallmeHongmaybe** [comment]
In my case I'd simply put the whole function in event listeners. 
Something like 
"""
onClick = {e => callback(e) }
"""

**Reply by CallmeHongmaybe** [reply]
And anyways thanks for making useCallback easy to digest.

**Comment by chiptus** [comment]
Thanks!!

**Comment by kishoreandra** [comment]
Great one dodo 🤩, Thanks for writing

**Comment by jamieday** [comment]
Really appreciate the tip! Good point that the ref may not be populated yet.

I do have a concern about the solution though - I remember reading in React docs that referential stability in useCallback is not a guarantee; the memo hooks are performance optimizations, but they’re not designed to semantically represent the fact that they’d return the same reference each time. So instead of using memoization, React suggests to use refs! So what that means here is that you’d probably want a ref-ref instead of a callback-ref 😅

The plot thickens, as I’m betting the same principle applies regarding the behavior of the ref prop - if that’s purely a performance optimization and not a semantic guarantee, then you shouldn’t rely on that behavior being stable either. So if that’s true you’d need to check that manually. Annnddd at this point we might want a custom hook. React is fun sometimes…

**Reply by TkDodo** [reply]
That's an interesting thought, thanks for pointing it out. I'm aware that `useMemo` has no semantic guarantee and that it's purely a perf optimization - I have a blogpost on the topic of _not_ using it to initialize values:
- https://tkdodo.eu/blog/use-state-for-one-time-initializations

`useCallback` doesn't have this section in the docs:

> You may rely on useMemo as a performance optimization, not as a semantic guarantee. In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without useMemo — and then add it to optimize performance.

However, I remember seeing somewhere else that `useCallback` is just `useMemo` where the return value is a function. But that doesn't necessarily mean that `useCallback` is actually _implemented_ with `useMemo`. This article basically asks the same question:

https://medium.com/trabe/can-we-rely-on-usecallback-as-a-semantic-guarantee-9329c17f40cf

The new beta docs for `useCallback` have not been written yet, so I'm curious to see how it will be phrased there: https://beta.reactjs.org/apis/usecallback

**Reply by gpichot** [reply]
I didn't know about `useMemo` not being a semantic guarantee. Very interesting to know. And didn't know about callback refs, thank you for this article @TkDodo !

`useCallback` does not seem to be implemented using `useMemo` on main branch, but in a similar manner. (Source https://github.com/facebook/react/blob/main/packages/react-reconciler/src/ReactFiberHooks.new.js#L1869-L1922 )

**Reply by jamieday** [reply]
This definitely seems like quite the rabbit hole… though I doubt React would change the memoization behavior without a big announcement anyway.

Thanks again @TkDodo for the article!

**Reply by dalvallana** [reply]
So, this would always work, right?
```jsx
const ref = React.useRef((node) => {
  node?.focus()
})

return <input ref={ref.current} defaultValue="Hello world" />

```

**Reply by TkDodo** [reply]
not sure what you mean with "always work". It will always run, on every render. Once with `null` and once with the newly rendered `node`. This might be fine, but for focussing as in the example, it's probably too often. That's why `useCallback` is needed.

**Reply by dalvallana** [reply]
> Luckily, React uses referential stability to check if the callback ref should be run or not. That means if we pass the same ref(erence, pun intended) to it, execution will be skipped

Shouldn't execution be skipped after the first render given that we're passing the same reference?

**Reply by TkDodo** [reply]
oh I think I totally misread your code because you actually put the function in a useRef. I'm not sure if it works, have you tried it? But yeah, I think it could actually work, as long as you don't have dependencies to the function where you'd actually need them to be re-created.

**Reply by TkDodo** [reply]
so, I just realised: if you don't have dependencies, you can also just move the function out of the component to make it always stable:

```
const ref = (node) => {
  node?.focus()
})

function Component() {
  return <input ref={ref} defaultValue="Hello world" />
}
```

**Comment by slashnot** [comment]
I wish I know this already. This would have solved many of my annoying problems in the past.

**Comment by xmd5a** [comment]
nice one! your articles are valuable; very interesting approach :)

**Comment by ghaoui** [comment]
Many many thanks

**Comment by jackblackCH** [comment]
As always, on point.

**Comment by mirik999** [comment]
Super, very useful solution. Thank you

**Comment by shataev** [comment]
Thanks a lot for this explanation!

**Comment by cagils** [comment]
Doesn't useLayoutEffect specifically for this use case?

**Reply by TkDodo** [reply]
no, not really. We are not layouting anything. We just want to make sure that the effect runs first, and as far as I could find out, `useEffect` orders are stable between the same component.

**Comment by samandar-boymurodov** [comment]
Nice article. Learnt about callback refs and why it is better to use it over ref objects in some cases. Thank you for your time and effort for making this!

**Comment by gavinthomas-valtech** [comment]
So using useLayoutEffect for focusing inputs is wrong?

https://blog.octo.com/why-you-should-use-uselayouteffect-when-tinkering-with-focus/

**Reply by TkDodo** [reply]
It's not what i would do :)

**Comment by gavinthomas-valtech** [comment]
Any ideas how we can type a heading that need focus
```
  const headingRef = useCallback((node: RefObject<HTMLElement>) => {
    node?.focus();
  }, []);
```
I get 'Property 'focus' does not exist on type 'RefObject<HTMLElement>'

or Property 'focus' does not exist on type 'RefObject<HTMLHeadingElement>'

**Reply by gavinthomas-valtech** [reply]
I'm using a component that states
  ```ref?: RefObject<HTMLHeadingElement>```

**Reply by gavinthomas-valtech** [reply]
although starting to wonder if in fact the heading can be focussable?

**Reply by TkDodo** [reply]
yeah I think the problem is that heading is not focusable :)

**Comment by hrachhovsepyanqless** [comment]
Would you use a ref callback to update a css variable also? Or would you rather use useLayoutEffect/useEffect?

```javascript
const setContainerOffset = (container) => {
  if (!container) return;

  const { y } = container.getBoundingClientRect();
  document.documentElement.style.setProperty("--offset", `${y}px`);
};


const MainPage = () => {
 return <div ref={setContainerOffset} />
}
```

**Reply by TkDodo** [reply]
I'd use a ref callback here, too, for pretty much the same reasons:
- it's less code
- it we pass the ref to a custom component, it might not have rendered immediately. The ref callback will be invoked when the component the ref is attached to actually mounted, not when the parent mounted.

**Reply by hrachhovsepyanqless** [reply]
Thank you!

**Comment by Seddy24** [comment]
This works awesome on initial render.  I also need to update the width on a resize as well.  Ideas on how to do that inside the useEffect watching for the resize?

```
  const [width, setWidth] = React.useState(0);
  
  const ref = React.useCallback(node => {
    if (node !== null) {
      setWidth(node.getBoundingClientRect().width)
    }
  }, [])

  useEffect(() => {
    function handleWindowResize() {
      //// need to set the width here on resize
      console.log('handleWindowResize')      
    }

    window.addEventListener('resize', handleWindowResize);

    return () => {
      window.removeEventListener('resize', handleWindowResize);
    };
  }, []);

**Reply by TkDodo** [reply]
No, I think you would still need to have a normal `ref` then. But you can combine it; the idea of the callback ref is that it always gets executed once when the component mounts, which is what you want. You can additionally store a ref:

```diff
  const [width, setWidth] = React.useState(0);
+ const ref = useRef()
  
  const ref = React.useCallback(node => {
    if (node !== null) {
+     ref.current = node
      setWidth(node.getBoundingClientRect().width)
    }
  }, [])

  useEffect(() => {
    function handleWindowResize() {
      //// need to set the width here on resize
      console.log('handleWindowResize')
+     if (ref.current) {
+       setWidth(ref.current.getBoundingClientRect().width)
+     }
    }

    window.addEventListener('resize', handleWindowResize);

    return () => {
      window.removeEventListener('resize', handleWindowResize);
    };
  }, []);
```

**Reply by Seddy24** [reply]
That makes sense and worked perfectly!

**Comment by AhmedBaset** [comment]
I hope I've read this before
thanks man

**Comment by mickykebe** [comment]
Great post. Thanks for explaining.

**Comment by polooner** [comment]
how could you give this a special ref type? like a type of HTMLDivElement for example?

**Reply by TkDodo** [reply]
```
const ref = React.useCallback((node: HtmlDivElement | null) => ...
```

**Comment by ashuvssut** [comment]
this is amazing. Thanks!!

**Comment by suryaashish-tyke** [comment]
I can do anything in the ref callback right?
like navigating to a page.

```tsx
const navigate = useNavigate()
<div ref={() => navigate("/redirect")}></div>
```

**Reply by TkDodo** [reply]
well yeah, technically, but I wouldn’t. In terms of being explicit, I would only use the callback ref when you actually need the `node` that is passed to it.

**Comment by nktnet1** [comment]
Great article!

With React 19 beta usage of the new compiler
- https://react.dev/learn/react-compiler#what-does-the-compiler-do

should anything here be revised?

**Reply by TkDodo** [reply]
I literally wrote a second article because of that:

- https://tkdodo.eu/blog/ref-callbacks-react-19-and-the-compiler

**Comment by CharlieBreval** [comment]
Your article have been so usefull, thank you very much

**Comment by pranavdhamanage-work** [comment]
This is fucking gold. Love it.

**Comment by stv8** [comment]
> Well, a tiny detail does: React will run this function after every render. So unless we are fine with focussing our input that often (which we are likely not), we have to tell React to only run this when we want to.

Where did you find this in the docs? I banged my head on a bug that I think was related to this last week but couldn't find the exact documentation I was looking for. Specifically I finally found out that changing the function to be stable fixed my issue (moved it outside of the render)

Curious if you have further docs you can point me to, thanks!

**Reply by TkDodo** [reply]
https://react.dev/reference/react-dom/components/common#ref-callback

> React will also call your ref callback whenever you pass a different ref callback. In the above example, (node) => { ... } is a different function on every render. When your component re-renders, the previous function will be called with null as the argument, and the next function will be called with the DOM node.

**Reply by stv8** [reply]
@TkDodo ah I swear I read that page 3 times, maybe I read it too fast as I was rushing to fix a bug 🙃  

Thanks for the link and the article, cheers!!
